modal/experimental.py

```diff
@@ -13,3 +31,81 @@ def get_local_input_concurrency():
     """Get the container's local input concurrency. Return 0 if the container is not running."""
 
     return _ContainerIOManager.get_input_concurrency()
+
+
+# START Experimental: Container Networking
+
+
+class _GroupedFunction(typing.Generic[P, ReturnType, OriginalReturnType], _Object, type_prefix="gf"):
+    def __init__(self, f: _Function, size: int):
+        self.f = synchronize_api(f)
+        self.size = size
+
+    def remote(self, *args: P.args, **kwargs: P.kwargs) -> List[ReturnType]:
+        """
+        Calls the function remotely, executing it with the given arguments and returning the execution's result.
+        """
+        worker_handles: List[FunctionCall] = []
+        with modal.Queue.ephemeral() as q:
+            for i in range(self.size):
+                handle = self.f.spawn(*args, **{**kwargs, "rank": i, "size": self.size, "q": q})
+                worker_handles.append(handle)
+        output: list[ReturnType] = []
+        for i, handle in enumerate(worker_handles):
+            output.append(handle.get())
+        return output
+
+    def local(self, *args: P.args, **kwargs: P.kwargs) -> ReturnType:
+        raise NotImplementedError("Grouped function cannot be run locally")
+
+    def spawn(self, *args: P.args, **kwargs: P.kwargs) -> ReturnType:
+        raise NotImplementedError("Grouped function cannot be spawned")
+
+
+def grouped(size=1):
+    def wrapper(raw_f: Callable[..., Any]) -> _PartialFunction:
+        if isinstance(raw_f, _Function):
+            raw_f = raw_f.get_raw_f()
+            raise InvalidError(
+                f"Applying decorators for {raw_f} in the wrong order!\nUsage:\n\n"
+                "@app.function()\n@modal.grouped()\ndef grouped_function():\n    ..."
+            )
+        raw_f = _networked(raw_f)
+        return _PartialFunction(raw_f, _PartialFunctionFlags.FUNCTION | _PartialFunctionFlags.GROUPED, group_size=size)
+
+    return wrapper
+
+
+def _networked(func):
+    def wrapper(*args, **kwargs):
+        import os
+
+        rank = kwargs.pop("rank", None)
+        size = kwargs.pop("size", None)
+        q = kwargs.pop("q", None)
+
+        if rank is None or size is None or q is None:
+            raise ValueError("This must be called using grouped decorator")
+
+        if not rank:
+            import socket
+
+            addr_info = socket.getaddrinfo("i6pn.modal.local", None, socket.AF_INET6)
+            # Extract IPv6 addresses from the results
+            ipv6_addresses = [
+                addr[4][0] for addr in addr_info if addr[1] == socket.SOCK_STREAM and "fdaa" in addr[4][0]
+            ]
+            main_ip = ipv6_addresses[0]
+            q.put_many([main_ip for _ in range(size)])
+        main_ip = q.get()
```

Comment:
I was wondering why you were using a Queue to share a single piece of information — I think the answer to my question might be that this will block until the main container has populated the queue with a main_ip for each replica to get? If that's right, it's smart! But maybe worth explaining in a comment.

Potentially we could (at some point) develop a new primitive to make this "share a single piece of information with multiple containers who can block until it's available" pattern more obvious.

Reply:
Yes that's intended behavior. We don't want containers to start until they have all synchronized with the main container's IP. Adding a comment